### PR TITLE
efi: Make use of the grub's prefix for detection

### PR DESCRIPTION
**Pull request by chrisccoulson**
Avoid relying on the signature in order to detect an Ubuntu production
grub binary for binaries without a SBAT section, as this breaks test cases
in snapd where grub is re-signed. Instead, obtain the prefix from the
"mods" section (which is set to "/EFI/ubuntu" in Ubuntu).

### CODE REVIEW

File: efi/grub.go [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+		if err := binary.Read(section, binary.LittleEndian, &info); err != nil {
+			return nil, fmt.Errorf("cannot obtain modules info: %w", err)
+		}
+		if info.Magic != grubModuleMagic {
+			return nil, errors.New("invalid module magic")
+		}
+		if info.Size < info.Offset {
+			return nil, errors.New("invalid modules size")
+		}
+		r = io.NewSectionReader(section, int64(info.Offset), int64(info.Size)-int64(info.Offset))
```
Comment (pedronis) [inline efi/grub.go:120]: this case is not tested

File: efi/image_rules_defs.go [hunk 2]
```diff
@@ -163,7 +163,7 @@ func makeMicrosoftUEFICASecureBootNamespaceRules() *secureBootNamespaceRules {
 				),
 				imageMatchesAll(
 					imageSectionExists("mods"),
-					imageSignedByOrganization("Canonical Ltd."),
+					grubHasPrefix("/EFI/ubuntu"),
```
[the remark this answers is not in the document]
Comment (chrisccoulson) [reply efi/image_rules_defs.go:166]: We do

File: efi/grub.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+	switch h.Machine() {
+	case pe.IMAGE_FILE_MACHINE_AMD64, pe.IMAGE_FILE_MACHINE_ARM64, pe.IMAGE_FILE_MACHINE_RISCV64:
+		var info grubModuleInfo64
+		if err := binary.Read(section, binary.LittleEndian, &info); err != nil {
+			return nil, fmt.Errorf("cannot obtain modules info: %w", err)
+		}
+		if info.Magic != grubModuleMagic {
+			return nil, errors.New("invalid modules magic")
+		}
+		if info.Size > math.MaxInt64 {
```
[the remark this answers is not in the document]
Comment (chrisccoulson) [reply efi/grub.go]: This was addressed with https://github.com/snapcore/secboot/pull/268/commits/421f90ae4cbbd93fb5ee8cbdc1051c74de8e0e00

**Review by pedronis — APPROVED** [review]
thank you

**Pull request merged**